circt/Dockerfile

```diff
@@ -0,0 +1,6 @@
+FROM alpine
+
+RUN apk --no-cache add build-base clang cmake git ninja python3
+
+COPY *.sh /tmp/
+RUN cd /tmp && /tmp/circt.sh
```

Comment:
Docker stores a temporary image for each `RUN` so it can very quickly replay steps which it's already cached. The problem with that is it cannot tell if there are changes from the outside world which affect steps. So it's *possible* for this Dockerfile to run from an older cached version instead of re-running this script. I don't think this would happen in the GH workflow since I don't think they cache images between runs, but this behavior may change in the future for efficiency's sake. (Pulling down Docker images for each of the builds which happen in docker containers is probably quite costly.)

I would suggest adding a git commit hash parameter to this Dockerfile indicating the circt commit you want to build, inlining the circt.sh script, and cloning the requested commit. In the workflow, you'd have to get the commit from circt/main and pass it into `docker build`

Reply:
Ah, cool. I didn't know this was how `RUN` worked (would've helped if I read the documentation... 😄). I can switch to the git commit hash approach.

There's probably a further optimization of making both the CIRCT and LLVM submodule hashes separate so that the latter can be cached if it didn't change. This is probably a future enhancement, though.

Reply:
yeah, I was considering suggesting that you also create an MLIR image that we could base this on, but that's probably unnecessary given that this is a nightly job.